### ISSUE TITLE
Particular box options don't work for `tcbincludepdf`

### ISSUE THREAD

**Issue by nignasi** [T-F-S/tcolorbox#128]
It seems to me that style options like `raster number x` or `raster column x` which work for `tcolorboxes` included into a raster, are not applied when the inner boxes are created with a `tcbincludepdf` command. If this command is replaced by a 

\foreach \i in {...}{\tcbincludegraphics[graphic options={page=\i}]{file to be included}} 

they are correctly applied.

**Comment by muzimuzhi** [comment]
Could you provide a complete example so others can directly work on it?

**Comment by muzimuzhi** [comment]
For a fix with minimal change, this should work

```diff
--- a/tex/latex/tcolorbox/tcbskins.code.tex
+++ b/tex/latex/tcolorbox/tcbskins.code.tex
@@ -1831,13 +1831,13 @@
 
 \newcommand{\tcbincludepdf}[2][]{%
   \begingroup%
-  \tcb@layer@pushup%
   \tcb@apply@box@options{enhanced,clip upper=true,valign=center,center upper,#1}%
   \appto\kvtcb@graphics@options{,page=\imagepage}%
   \tcb@include@graphics@preparefile{#2}%
   \tcb@set@pdfpages\pdfpages{\imagepath}%
   \foreach \imagepage in \kvtcb@graphics@pages%
   {%
+    \tcb@layer@pushup%
     \tcb@include@graphics%
   }%
   \endgroup%
```

In short, `\tcb@layer@pushup` should be executed before every tcolorbox inside raster.

A longer explanation: 
 - `\tcb@layer@pushup` calls style `every box on layer n` _and_ clear it.
 - `every box on layer n` contains raster style options like `raster number x`.
 -  To make `raster number x` work as expected, it should be executed per tcolorbox inside raster.
 - In `\tcbincludepdf`, every `\tcb@include@graphics` expands to a local tcolorbox (thanks to `\foreach`), but currently `\tcb@layer@pushup` is called only once. Hence the raster styles applied to the first inner tcolorbox wrongly affect every tcolorboxs in `\tcbincludepdf`.

------
A simplified example to reproduce and test this problem:
```tex
\documentclass{article}
\usepackage[margin=10pt]{geometry}
\usepackage[skins,raster]{tcolorbox}

\begin{document}
\tcbset{
  raster columns=3,
  raster number 1/.style={colback=blue!10}
}

\begin{tcbraster}[
  title=\arabic{tcblayer},
]
  \foreach \i in {1,...,3} {%
    \tcbincludegraphics[graphics options={page=\i}]{2.pdf}}
\end{tcbraster}

\begin{tcbraster}[
  title=\arabic{tcblayer},
  graphics pages={1,...,3}
]
  \tcbincludepdf{2.pdf}
\end{tcbraster}
\end{document}
```
where `example-pages.pdf` is produced by
```tex
% example-pages
\documentclass[a4paper, landscape]{article}
\usepackage{graphicx}
\usepackage{pgffor}
\pagestyle{empty}

\begin{document}
\centering
\foreach \i in {1,...,6} {
  \scalebox{10}{page \i}\newpage
}
\end{document}
```

**Comment by muzimuzhi** [comment]
In case user has changed `every box on layer n`, perhaps inside `tcbraster ` environment, using `pgfkeys` handler `/.append code` is better than `/.code`
https://github.com/T-F-S/tcolorbox/blob/cc675797483e436768e59cc21e3d9fc4d1823616/tex/latex/tcolorbox/tcbraster.code.tex#L136

**Comment by T-F-S** [comment]
I agree to all explanations. For the patch of `\tcbincludepdf` I have a slight variation:

```latex
\renewcommand{\tcbincludepdf}[2][]{%
  \tcb@include@graphics@preparefile{#2}%
  \tcb@set@pdfpages\pdfpages{\imagepath}%
  \foreach \imagepage in \kvtcb@graphics@pages%
  {%
    \begingroup%
    \tcb@layer@pushup%
    \tcb@apply@box@options{enhanced,clip upper=true,valign=center,center upper,#1}%
    \appto\kvtcb@graphics@options{,page=\imagepage}%
    \tcb@include@graphics%
    \endgroup%
  }%
}
```

I have not tested the patch outside the MWE yet, but if nothing comes up, I will change the implementation accordingly.

Also, the suggestion for appending instead replacing the layer code inside the rasters seems reasonable to me.

Thank you.

**Comment by nignasi** [comment]
First of all, thank you for your fast reaction.

As [mizimuzhi ](https://github.com/muzimuzhi) commented, I was trying to provide an answer to [this question](https://tex.stackexchange.com/q/577339/1952) when I found what I think it's a bug.

Although I'm sure that you will find the best solution for this mistake, I won't use it on my answer until it would be officially  incorporated into the package.

**Comment by muzimuzhi** [comment]
@T-F-S I think my patch is sufficient since each time the loop body of `\foreach` is executed, it is executed in a TeX group. See the doc for `\foreach` in `pgfmanual`, sec. 88, [the paragraph just before bold text "Syntax for the commands."](Syntax for the commands.).

**Comment by T-F-S** [comment]
Fixed with version 4.50 (2021/05/21)
